File: language/compiler/src/com/kubadziworski/parsing/visitor/statement/AssignmentStatementVisitor.java
package com.kubadziworski.parsing.visitor.statement;

import jiva.parser.JivaBaseVisitor;
import jiva.parser.JivaParser;
import com.kubadziworski.domain.node.expression.Expression;
import com.kubadziworski.domain.node.statement.Assignment;
import com.kubadziworski.parsing.visitor.expression.ExpressionVisitor;
import org.antlr.v4.runtime.misc.NotNull;

public class AssignmentStatementVisitor extends JivaBaseVisitor<Assignment> {
    private final ExpressionVisitor expressionVisitor;

    public AssignmentStatementVisitor(ExpressionVisitor expressionVisitor) {
        this.expressionVisitor = expressionVisitor;
    }

    @Override
    public Assignment visitAssignment(@NotNull JivaParser.AssignmentContext ctx) {
        JivaParser.ExpressionContext expressionCtx = ctx.expression();
        Expression expression = expressionCtx.accept(expressionVisitor);
        String varName = ctx.name().getText();
        return new Assignment(varName, expression);
    }
}

File: language/compiler/src/parsing/visitor/statement/AssignmentStatementVisitor.kt
package jiva.parsing.visitor.statement

import jiva.domain.node.statement.Assignment
import jiva.parsing.visitor.expression.ExpressionVisitor
import jiva.parser.JivaBaseVisitor
import jiva.parser.JivaParser.AssignmentContext

class AssignmentStatementVisitor(private val expressionVisitor: ExpressionVisitor) : JivaBaseVisitor<Assignment?>() {

    override fun visitAssignment(ctx: AssignmentContext): Assignment {
        val expressionCtx = ctx.expression()
        val expression = expressionCtx.accept(expressionVisitor)
        val varName = ctx.name().text
        return Assignment(varName, expression!!)
    }

}
